File: frontend/src/api/profileBranding.ts
import { api } from "./client";
import { mediaUrl } from "../lib/media-url";

export interface PublicBrandingOut {
  app_name: string;
  topbar_logo_url: string | null;
  institution_logo_url: string;
  institution_logo_enabled: boolean;
  institution_logo_display_size_px: number;
  login_banner_url: string | null;
  login_banner_enabled: boolean;
  updated_at: string;
}

export interface UserProfileOut {
  id: number;
  role: string;
  full_name: string;
  email: string;
  login?: string;
  group_id?: number;
  avatar_url?: string;
}

export interface ProfileMeOut {
  user: UserProfileOut;
  role_details: Record<string, any>;
}

export interface UserAvatarOut {
  avatar_id: number;
  url: string;
  is_active: boolean;
  created_at: string;
}

export interface UserAvatarsListOut {
  items: UserAvatarOut[];
  limit: number;
}

export interface AuthSessionOut {
  auth_session_id: string;
  is_current: boolean;
  client_kind: string;
  device_name?: string;
  os_name?: string;
  os_version?: string;
  browser_name?: string;
  browser_version?: string;
  app_version?: string;
  location_label: string;
  ip_addr_masked: string;
  created_at: string;
  last_seen_at: string;
  expires_at: string;
}

export interface AuthSessionsListOut {
  items: AuthSessionOut[];
}

export interface BrandingAssetOut {
  asset_id: number;
  asset_kind: string;
  url: string;
  content_type: string;
  size_bytes: number;
  width_px: number;
  height_px: number;
  display_width_px?: number;
  display_height_px?: number;
  original_name?: string;
  created_at: string;
}

export interface AdminBrandingOut {
  app_name: string;
  institution_logo_enabled: boolean;
  institution_logo_display_size_px: number;
  login_banner_enabled: boolean;
  topbar_logo_asset_id?: number;
  institution_logo_asset_id?: number;
  login_banner_asset_id?: number;
  active_assets: Record<string, BrandingAssetOut | null>;
  recent_assets: Record<string, BrandingAssetOut[]>;
}

export const profileBrandingApi = {
  getPublicBranding: () =>
    api.get<PublicBrandingOut>("/api/v2/profile-branding/public/branding").then((res) => normalizePublicBranding(res.data)),

  getProfileMe: () =>
    api.get<ProfileMeOut>("/api/v2/profile-branding/me").then((res) => normalizeProfile(res.data)),

  updateProfileMe: (patch: { login?: string; email?: string }) =>
    api.patch<ProfileMeOut>("/api/v2/profile-branding/me", patch).then((res) => normalizeProfile(res.data)),

  getAvatars: () =>
    api.get<UserAvatarsListOut>("/api/v2/profile-branding/avatars").then((res) => ({
      ...res.data,
      items: res.data.items.map(normalizeAvatar),
    })),

  uploadAvatar: (file: File, cropJson?: string) => {
    const fd = new FormData();
    fd.append("file", file);
    if (cropJson) {
      fd.append("crop_json", cropJson);
    }
    return api.post<UserAvatarOut>("/api/v2/profile-branding/avatars", fd, {
      headers: { "Content-Type": "multipart/form-data" },
    }).then((res) => normalizeAvatar(res.data));
  },

  activateAvatar: (avatarId: number) =>
    api.post<UserAvatarOut>(`/api/v2/profile-branding/avatars/${avatarId}/activate`).then((res) => normalizeAvatar(res.data)),

  deleteAvatar: (avatarId: number) =>
    api.delete(`/api/v2/profile-branding/avatars/${avatarId}`),

  getSessions: () =>
    api.get<AuthSessionsListOut>("/api/v2/profile-branding/sessions").then((res) => res.data),

  revokeSession: (sessionId: string) =>
    api.delete(`/api/v2/profile-branding/sessions/${sessionId}`),

  revokeAllSessions: (includeCurrent = false) =>
    api.delete<{ revoked_count: number; current_revoked: boolean }>(
      `/api/v2/profile-branding/sessions?include_current=${includeCurrent}`
    ).then((res) => res.data),

  getAdminBranding: () =>
    api.get<AdminBrandingOut>("/api/v2/profile-branding/admin/branding").then((res) => normalizeAdminBranding(res.data)),

  updateAdminBranding: (settings: Partial<PublicBrandingOut & {
    topbar_logo_asset_id: number | null;
    institution_logo_asset_id: number | null;
    login_banner_asset_id: number | null;
  }>) =>
    api.patch<PublicBrandingOut>("/api/v2/profile-branding/admin/branding", settings).then((res) => normalizePublicBranding(res.data)),

  uploadBrandingAsset: (kind: string, file: File) => {
    const fd = new FormData();
    fd.append("file", file);
    return api.post<BrandingAssetOut>(`/api/v2/profile-branding/admin/branding/assets/${kind}`, fd, {
      headers: { "Content-Type": "multipart/form-data" },
    }).then((res) => normalizeAsset(res.data));
  },

  deleteBrandingAsset: (assetId: number) =>
    api.delete(`/api/v2/profile-branding/admin/branding/assets/${assetId}`),
};

function normalizePublicBranding(data: PublicBrandingOut): PublicBrandingOut {
  return {
    ...data,
    topbar_logo_url: mediaUrl(data.topbar_logo_url),
    institution_logo_url: mediaUrl(data.institution_logo_url) ?? "",
    institution_logo_enabled: data.institution_logo_enabled ?? true,
    login_banner_url: mediaUrl(data.login_banner_url),
  };
}

function normalizeProfile(data: ProfileMeOut): ProfileMeOut {
  return {
    ...data,
    user: {
      ...data.user,
      avatar_url: mediaUrl(data.user.avatar_url) ?? undefined,
    },
  };
}

function normalizeAvatar(data: UserAvatarOut): UserAvatarOut {
  return {
    ...data,
    url: mediaUrl(data.url) ?? data.url,
  };
}

function normalizeAsset(data: BrandingAssetOut): BrandingAssetOut {
  return {
    ...data,
    url: mediaUrl(data.url) ?? data.url,
  };
}

function normalizeAdminBranding(data: AdminBrandingOut): AdminBrandingOut {
  const active_assets = Object.fromEntries(
    Object.entries(data.active_assets).map(([kind, asset]) => [kind, asset ? normalizeAsset(asset) : null]),
  );
  const recent_assets = Object.fromEntries(
    Object.entries(data.recent_assets).map(([kind, assets]) => [kind, assets.map(normalizeAsset)]),
  );
  return {
    ...data,
    institution_logo_enabled: data.institution_logo_enabled ?? true,
    active_assets,
    recent_assets,
  };
}

File: backend/app/schemas/profile_branding.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional, Dict
from pydantic import BaseModel, EmailStr, Field
import uuid

class ClientInfo(BaseModel):
    client_kind: str = "web"
    device_name: Optional[str] = None
    os_name: Optional[str] = None
    os_version: Optional[str] = None
    browser_name: Optional[str] = None
    browser_version: Optional[str] = None
    app_version: Optional[str] = None
    timezone: Optional[str] = None
    locale: Optional[str] = None

class UserProfileOut(BaseModel):
    id: int
    role: str
    full_name: str
    email: EmailStr
    login: Optional[str] = None
    group_id: Optional[int] = None
    avatar_url: Optional[str] = None

class ProfileMeOut(BaseModel):
    user: UserProfileOut
    role_details: dict = {}

class UserAvatarOut(BaseModel):
    avatar_id: int
    url: str
    is_active: bool
    created_at: datetime

class UserAvatarsListOut(BaseModel):
    items: List[UserAvatarOut]
    limit: int = 12

class AuthSessionOut(BaseModel):
    auth_session_id: str
    is_current: bool
    client_kind: str
    device_name: Optional[str] = None
    os_name: Optional[str] = None
    os_version: Optional[str] = None
    browser_name: Optional[str] = None
    browser_version: Optional[str] = None
    app_version: Optional[str] = None
    location_label: str
    ip_addr_masked: str
    created_at: datetime
    last_seen_at: datetime
    expires_at: datetime

class AuthSessionsListOut(BaseModel):
    items: List[AuthSessionOut]

class PublicBrandingOut(BaseModel):
    app_name: str = "Верифика"
    topbar_logo_url: Optional[str] = None
    institution_logo_url: Optional[str] = "/collegelogo.jpg"
    institution_logo_enabled: bool = True
    institution_logo_display_size_px: int = 56
    login_banner_url: Optional[str] = None
    login_banner_enabled: bool = False
    updated_at: datetime

class BrandingAssetOut(BaseModel):
    asset_id: int
    asset_kind: str
    url: str
    content_type: str
    size_bytes: int
    width_px: int
    height_px: int
    display_width_px: Optional[int] = None
    display_height_px: Optional[int] = None
    original_name: Optional[str] = None
    created_at: datetime

class AdminBrandingOut(BaseModel):
    app_name: str
    institution_logo_enabled: bool
    institution_logo_display_size_px: int
    login_banner_enabled: bool
    topbar_logo_asset_id: Optional[int] = None
    institution_logo_asset_id: Optional[int] = None
    login_banner_asset_id: Optional[int] = None
    active_assets: Dict[str, Optional[BrandingAssetOut]] = {}
    recent_assets: Dict[str, List[BrandingAssetOut]] = {}
